extension/src/popup/helpers/sep24.ts

```diff
@@ -0,0 +1,180 @@
+import StellarSdk, { Account, StellarTomlResolver } from "stellar-sdk";
+
+import { NetworkDetails } from "@shared/constants/stellar";
+
+import { emitMetric } from "helpers/metrics";
+import { xlmToStroop } from "helpers/stellar";
+import { AppDispatch } from "popup/App";
+
+import { METRIC_NAMES } from "popup/constants/metricsNames";
+import { Sep24Status } from "popup/constants/sep24";
+import {
+  startHwSign,
+  signFreighterTransaction,
+  submitFreighterTransaction,
+  storeSep24Status,
+  clearSep24Data,
+} from "popup/ducks/transactionSubmission";
+
+export const getAuthToken = async ({
+  signedXdr,
+  sep10Url,
+}: {
+  signedXdr: string;
+  sep10Url: string;
+}) => {
+  const authRes = await fetch(sep10Url, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ transaction: signedXdr }),
+  });
+  const resJson = await authRes.json();
+  return resJson.token;
+};
+
+export const getAnchorSep24Data = async ({
+  anchorDomain,
+  publicKey,
+}: {
+  anchorDomain: string;
+  publicKey: string;
+}) => {
+  const tomlRes = await StellarTomlResolver.resolve(anchorDomain);
+  const anchorSigningKey = tomlRes.SIGNING_KEY;
+
+  const auth = await fetch(`${tomlRes.WEB_AUTH_ENDPOINT}?account=${publicKey}`);
+  const authJson = await auth.json();
+  const challengeTx = authJson.transaction;
+
+  const { tx } = StellarSdk.Utils.readChallengeTx(
+    challengeTx,
+    anchorSigningKey,
+    authJson.network_passphrase,
+    anchorDomain,
+    anchorDomain,
+  );
+  const transactionXDR = tx.toXDR();
+
+  return {
+    sep10Url: tomlRes.WEB_AUTH_ENDPOINT,
+    sep24Url: tomlRes.TRANSFER_SERVER_SEP0024,
+    challengeTx: transactionXDR,
+    networkPassphrase: authJson.network_passphrase,
+  };
+};
+
+export const startSep24Polling = async ({
+  dispatch,
+  sep24Url,
+  txId,
+  token,
+  status,
+}: {
+  dispatch: AppDispatch;
+  sep24Url: string;
+  txId: string;
+  token: string;
+  status: string;
+}) => {
+  let currentStatus = status as Sep24Status;
+  const endStatuses = [
+    Sep24Status.INCOMPLETE,
+    Sep24Status.PENDING_TRUST,
+    Sep24Status.ERROR,
+  ];
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
```

Comment:
assuming `true` is placeholder for now?

Reply:
actually I thought of keeping it as is. The `startSep24Polling` should only be called if there's sep24 data in the storage, meaning there's a sep24 todo that needs to be resolved with the anchor. So the only polling ending case is the anchor giving use one of the "endStatuses" (or COMPLETED).

To be safe maybe we only ping a max number of times? Like 10? wdyt?